app/services/UIBuilder.scala

```diff
@@ -28,27 +28,35 @@ import scala.annotation.tailrec
 class UIBuilder {
   val logger: Logger = Logger(getClass)
 
-  def fromStanzas(url: String, stanzas: Seq[Stanza], formData: Option[FormData] = None)(implicit stanzaIdToUrlMap: Map[String, String]): Page =
-    Page(url,
-         stanzas.foldLeft(Seq[UIComponent]()) {(acc, stanza) =>
-           stanza match {
-             case i: Instruction => acc :+ fromInstruction(i)
-             case ig: InstructionGroup => acc :+ fromInstructionGroup(ig)
-             case c: Callout => acc ++ fromCallout(c, formData)
-             case rg: RowGroup => acc :+ fromRowGroup(rg)
-             case in: OcelotInput => Seq(fromInput(in, formData, acc))
-             case q: OcelotQuestion => Seq(fromQuestion(q, acc))
-             case _ => acc
-           }
-         })
-
-  private def fromRowGroup(rg: RowGroup)(implicit stanzaIdToUrlMap: Map[String, String]): UIComponent = {
-    val rowLength = rg.group.map(_.cells.length).max
-    SummaryList(rg.group.map{row =>
-      SummaryRow((row.cells ++ Seq.fill(rowLength - row.cells.size)(Phrase())).map{phrase =>
-        SummaryCell(TextBuilder.fromPhrase(phrase))
-      })
-    })
+  def buildPage(url: String, stanzas: Seq[VisualStanza], formData: Option[FormData] = None)
+               (implicit stanzaIdToUrlMap: Map[String, String]): Page =
+    Page(url, fromStanzas(stackStanzas(stanzas, Nil), Nil, formData))
+
+  @tailrec
+  private def fromStanzas(stanzas: Seq[VisualStanza], acc: Seq[UIComponent], formData: Option[FormData])
```

Comment:
IntelliJ is reporting a cyclomatic complexity problem.

Reply:
arghh !!!!